Zend/zend_compile.c

```diff
@@ -2867,8 +2865,33 @@ static void zend_compile_list_assign(
 
 		zend_verify_list_assign_target(var_ast, old_style);
 
-		zend_emit_op(&fetch_result, ZEND_FETCH_LIST, expr_node, &dim_node);
-		zend_emit_assign_znode(var_ast, &fetch_result);
+		opline = zend_emit_op(&fetch_result, ZEND_FETCH_LIST, expr_node, &dim_node);
+		if (elem_ast->attr) {
+			opline->extended_value = ZEND_LIST_MAKE_WRITABLE;
+			prev_op = (opline - 1);
+			do {
+				if (prev_op->opcode == ZEND_FETCH_LIST) {
+					if (prev_op->op1.var == opline->op1.var) {
+						continue;
+					} else if (prev_op->result.var == opline->op1.var) {
+						if (prev_op->extended_value == ZEND_LIST_MAKE_WRITABLE) {
+							break;
+						}
+						prev_op->extended_value = ZEND_LIST_MAKE_WRITABLE;
+						opline = prev_op;
+						continue;
+					}
+					break;
+				} else if (prev_op->opcode == ZEND_ASSIGN_REF || prev_op->opcode == ZEND_ASSIGN) {
```

Comment:
Instead of fixing up opcodes after the fact you can do an up-front pass through the list() ast and add a flag for all list()s that recursively contain a reference. Then you can directly emit the opcode with the correct flag.